server/src/main/java/org/elasticsearch/index/IndexVersion.java

```diff
@@ -115,14 +115,19 @@ private static IndexVersion registerIndexVersion(int id, Version luceneVersion,
     public static final IndexVersion V_8_9_0 = registerIndexVersion(8_09_00_99, Version.LUCENE_9_7_0, "32f6dbab-cc24-4f5b-87b5-015a848480d9");
     public static final IndexVersion V_8_9_1 = registerIndexVersion(8_09_01_99, Version.LUCENE_9_7_0, "955a80ac-f70c-40a5-9399-1d8a1e5d342d");
     public static final IndexVersion V_8_10_0 = registerIndexVersion(8_10_00_99, Version.LUCENE_9_7_0, "2e107286-12ad-4c51-9a6f-f8943663b6e7");
-    public static final IndexVersion V_8_11_0 = registerIndexVersion(8_11_00_99, Version.LUCENE_9_8_0, "f08382c0-06ab-41f4-a56a-cf5397275627");
+    public static final IndexVersion V_8_11_0 = registerIndexVersion(8_11_00_99, Version.LUCENE_9_7_0, "f08382c0-06ab-41f4-a56a-cf5397275627");
 
     /*
      * READ THE COMMENT BELOW THIS BLOCK OF DECLARATIONS BEFORE ADDING NEW INDEX VERSIONS
      * Detached index versions added below here.
      */
-    public static final IndexVersion V_8_500_000 = registerIndexVersion(8_500_000, Version.LUCENE_9_8_0, "bf656f5e-5808-4eee-bf8a-e2bf6736ff55");
-    public static final IndexVersion V_8_500_001 = registerIndexVersion(8_500_001, Version.LUCENE_9_8_0, "45045a5a-fc57-4462-89f6-6bc04cda6015");
+    // A synthetic version that is in the future and is on the next Lucene minor version
+    // It needs to be on the same major as the actual current version for backward compatibility assumptions to be correct
+    // Keep it above other versions so that adding new versions in main doesn't cause merge conflicts when merging into the lucene_snapshot
+    // branch
+    public static final IndexVersion UPGRADE_TO_NEXT_LUCENE_VERSION = registerIndexVersion(8_999_999, Version.LATEST, "ee5ab2e6-4d8f-11ee-be56-0242ac120002");
```

Comment:
I don't think we should use a fake version number because that requires carefully changing at the time this branch is forked to create a merge commit for the final Lucene release. My suggestion was to use a bump from the latest, whenever Lucene is updated.

So currently we are at IV 8500001, let's bump that to 8500010. This way any IV bumps on main will still be less than this number. I think we should also name the IV constant eg LUCENE_97_UPGRADE.

I suspect we will want to have a test in the future to verify these constants are in increasing order, so I would also move this constant down. I think if you move it below the "stop" comment, there won't be merge conflicts. Then when creating the branch for a release version of Lucene, the constant just needs to be moved up with the rest of them.

Reply:
One thing I liked about this fake version number was that it would still work when the version gets bumped on the Lucene side by using Version.LATEST, but I'm slightly reluctant to do it on a true version number as it would be easy to miss changing it back to the actual version number at merge time, do you have thoughts on this? I guess that not using `LATEST` is not the end of the world as the build would only break every 2 months or so because of the version bump.

I'm also curious as to whether you would change the version number at merge time to keep version numbers compact, or whether you would accept that there might be gaps? E.g. to take your example, assuming the current version is `8_500_007` when merging to main, would you update the version number to `8_500_008` or keep it `8_500_010`? Also I'm unclear as whether the version number would remain called `LUCENE_9_8_UPGRADE` or whether you would rename it to `V_8_500_010`?

Reply:
> change the version number at merge time to keep version numbers compact, or whether you would accept that there might be gaps?

gaps are fine, we will just need to think about them in mapping out the version number format (like we are doing for TransportVersion). I suspect it will look very similar. Doing a bump of 10 seems fine to me since we would then have a range of 1k for a major version (not a hard limit), and other additional IndexVersion changes occur within that 10, so we really aren't exhausting the space that much faster.

>  I'm unclear as whether the version number would remain called LUCENE_9_8_UPGRADE or whether you would rename it to V_8_500_010?

I would keep it as the meaningful name. We are switching to meaningful constant names for TransportVersion, and will do the same for IndexVersion, we just haven't gotten to renaming them yet.

Reply:
re using Version.LATEST or an explicit version, IMO an explicit version is better since no changes are needed when the branch with the release version is cut. In particular all the uses of the constant would stay the same, rather than needing to adjust every use of a constant name.

Bumping to a new snapshot version on this branch (eg 9.8-xxx to 9.9-xxx) would still require manual intervention right? In which case I think adjusting the constant can be part of that manual process.